File: common/src/main/java/com/example/interceptors/HeaderServerInterceptor.java
package com.example.interceptors;

import io.grpc.Context;
import io.grpc.Contexts;
import io.grpc.ForwardingServerCall.SimpleForwardingServerCall;
import io.grpc.Metadata;
import io.grpc.ServerCall;
import io.grpc.ServerCallHandler;
import io.grpc.ServerInterceptor;

import static com.example.interceptors.Headers.REQUEST_ID_KEY;
import static com.example.utils.LoggingUtils.log;

public class HeaderServerInterceptor implements ServerInterceptor {
    @Override
    public <ReqT, RespT> ServerCall.Listener<ReqT> interceptCall(
        ServerCall<ReqT, RespT> call,
        Metadata requestHeaders,
        ServerCallHandler<ReqT, RespT> next
    ) {
        log("headers received from client: %s%n", requestHeaders);

        var requestId = requestHeaders.get(REQUEST_ID_KEY);
        if (requestId == null) {
            return next.startCall(call, requestHeaders);
        }

        var context = Context.current().withValue(Headers.REQUEST_ID_CTX_KEY, requestId);
        return Contexts.interceptCall(context, new SimpleForwardingServerCall<>(call) {
            @Override
            public void sendHeaders(Metadata responseHeaders) {
                responseHeaders.put(REQUEST_ID_KEY, requestId);
                super.sendHeaders(responseHeaders);
            }
        }, requestHeaders, next);
    }
}

File: demo-grpc-proto/src/test/kotlin/com/example/util/HeaderServerInterceptor.kt
package com.example.util

import io.grpc.ForwardingServerCall.SimpleForwardingServerCall
import io.grpc.Metadata
import io.grpc.ServerCall
import io.grpc.ServerCallHandler
import io.grpc.ServerInterceptor
import java.util.*

class HeaderServerInterceptor : ServerInterceptor {
    private val key = Metadata.Key.of("response_id", Metadata.ASCII_STRING_MARSHALLER)

    override fun <ReqT : Any?, RespT : Any?> interceptCall(
        call: ServerCall<ReqT, RespT>?,
        requestHeaders: Metadata?,
        next: ServerCallHandler<ReqT, RespT>?
    ): ServerCall.Listener<ReqT> {
        println("[${Thread.currentThread()}] header received from client: $requestHeaders")
        return next!!.startCall(object : SimpleForwardingServerCall<ReqT, RespT>(call) {
            override fun sendHeaders(responseHeaders: Metadata) {
                responseHeaders.put(key, UUID.randomUUID().toString())
                super.sendHeaders(responseHeaders)
            }
        }, requestHeaders)
    }
}
